### PR TITLE
1749 admin styles leaking in gdocs preview iframe

### PR DESCRIPTION
**Pull request by mlbrgl**
**Summary of features:**

1. prevent admin styles leaking into the preview being embedding it into a regular "src" iframe
2. Reveal block errors on live site (rationale similar to highlighting "missing data" on charts)
3. expose articles drafts on a new public (unlisted) URL (http://localhost/gdocs/[ID]/preview), outside of the admin area, to allow external contributors to read article drafts in-situ. This means that if an external contributor is given access to both the external preview URL and the google doc, they will be able to edit and preview the result.
4. Suggestions don't show in the preview anymore

_See thread below for back and forth between technical solutions leading the choice made in 1._

**Deploy steps:**
- [ ] update nginx configuration to proxy `/gdocs` paths
```
location ~ ^/(admin|gdocs)/ {
    try_files $uri @nodeAdmin;
    add_header X-Robots-Tag noindex;
  }
sudo service nginx reload
```
- [ ] reload nginx `sudo service nginx reload`
- [ ] deploy as usual

### ***
_Original text below_

This isolate the preview in its own production-like iframe, so no admin styles are included.

See https://github.com/owid/owid-grapher/issues/1749#issuecomment-1342863046 for a previous attempt using a shadow DOM.

This now also wraps runtime errors in error boundaries on the live site. Previously, the error boundary would only be wrapping blocks when the debug context was set, which only happened for previews.

On the live site, errors in blocks are not as verbose as in debug mode (automatically set for previews), but are still visible so that missing data / charts is not hidden away, and the narrative still holds.

With "debug mode" (admin previews):
Screenshot 2023-01-08 at 18 09 10

Without "debug mode" (live site):
Screenshot 2023-01-08 at 18 08 25

TODO
- [x] rename view / preview
- [x] review debug context setup
- [x] restore auto-refresh?

### CODE REVIEW

**Comment by mlbrgl** [comment]
The goal here is to isolate the article preview from the admin styles.

For that purpose, I used an iframe, which renders an OwidArticlePage at its own URL.
In that configuration, the article rendered in the preview goes through the same rendering pipeline: an OwidArticlePage rendered server-side, then hydrated on the client. This gives us the isolation we needed and no leaking styles, in addition to the confidence that any possible rendering edge case would be captured and surfaced in the preview.

This solution also comes with a set of drawbacks:
- the Google API is hit twice for every refresh: once for the admin interface, and once for the iframe
- refreshing an iframe is visible to the user (flicker)
- refreshing an iframe scrolls its content back to the top

As I was addressing the last two items, I came across a simpler solution to our initial problem: rather than refreshing the iframe itself, we could instead re-render its content through a portal. This solves all three issues mentioned above, while still giving us complete style isolation.

The disadvantage of this solution is that we are going back to the OwidArticle being rendered client-side only, and we have to sticth together a client-side version of the OwidArticlePage. This makes the preview goes through a different rendering pipeline that the actual production article, which could hide potential problems until they hit production.

Even if it is not as straightforward, we can still keep the SSR version around at its /draft URL, as a way to check periodically when changes to the rendering pipeline are made.
As a silver lining, this /draft page could also be used as private previews of articles, when requesting feedback.

For now, the solution looks like:
- render the previewed OwidArticle client-side (no conceptual change), by embedding it in an iframe through a portal. Inject stylesheet imports and scripts.
- expose articles drafts on a new public (unlisted) URL, outside of the admin area

**Comment by mlbrgl** [comment]
I am a bit ambivalent about the solution outlined in this PR. 

We're effectively "paying" for the single request to the Google docs API by having an alternate rendering path for previews, which preserves scrolling position and doesn't flicker on refresh.

With the SSR solution, each change triggers two requests but we keep a simpler rendering pipeline (once debugging has been taken into account). Rendering is more brute-force here, and akin to a full page refresh (at the level of the preview iframe), which means we'd need to take care of authoring UX (scrolling position, flicker).

Let's see if that trade makes sense in the longer term.

**Comment by mlbrgl** [comment]
The straw(s) that broke last week camel's back:
- clicking on links within the iframe triggers "russian dolls":
Screenshot 2023-01-16 at 15 15 58

- visual bugs on charts UI
Screenshot 2023-01-16 at 15 15 25
Screenshot 2023-01-16 at 15 15 35

All of these might be fixable but until I can regain confidence in the preview final's output and rendering, I'll go back to the brute force solution of the "full refresh" iframe (which will only refresh and hit the google api if the first request detected another revision being pulled in).

**Comment by ikesau** [comment]
Hey @mlbrgl - just wondering if you looked into CSS layers at all to try and solve this? 

I've only [tinkered with them in a codepen](https://codepen.io/ikesaunders/pen/mdjMvww?editors=1100) but could we possibly have an admin layer at the base, then a reset layer, then our site styles?

**Comment by mlbrgl** [comment]
no I haven't, thanks for bringing that up to my attention @ikesau. 

I'm quite above what I had budgeted for this already so I won't look into it now, but that might be worth investigating for a next iteration.

One thing that puts my mind at ease with the iframe solution is how close it is guaranteed to be to the final product. But yes, it is a compromise.

@danyx23 tagging you since you also recently worked on embedding through iframe, they might be some synergies there. You also worked on error handling.

**Comment by mlbrgl** [comment]
Also, suggestions don't show in previews anymore

### Before
Screenshot 2023-01-16 at 14 35 34

### After
Screenshot 2023-01-16 at 14 39 43

File: adminSiteServer/app.tsx [hunk 1]
```diff
@@ -27,6 +28,9 @@ import { renderToHtmlPage } from "../serverUtils/serverUtil.js"
 import { publicApiRouter } from "./publicApiRouter.js"
 import { mockSiteRouter } from "./mockSiteRouter.js"
 import { GIT_CMS_DIR } from "../gitCms/GitCmsConstants.js"
+import { GdocsContentSource } from "@ourworldindata/utils/dist/owidTypes.js"
```
Comment (danyx23) [inline adminSiteServer/app.tsx:31]: Better to not do a deep import here I think
```suggestion
import { GdocsContentSource } from "@ourworldindata/utils"
```

**Review by danyx23 — COMMENTED** [review]
This looks good to me and would be very useful for the backporting work that I am doing atm. Is this deployed somewhere? Because this touches a few places of how our main scripts are run, what parameters are required etc I would like to run through the fully baked site on staging with these changes just to make sure everything is fine. Otherwise this could be merged from my side.

File: adminSiteClient/GdocsPreviewPage.tsx [hunk 2]
```diff
@@ -233,9 +234,46 @@ export const GdocsPreviewPage = ({ match, history }: GdocsMatchProps) => {
...
+                    authoring experience at the moment (content flashes and scrolling position
+                    resets on every change)
+                */}
+                <iframe
+                    src={`/gdocs/${gdoc.id}/preview#owid-article-root`}
+                    style={{ width: "100%", height: "inherit", border: "none" }}
+                    key={gdoc.revisionId}
+                />
+
+                {/* Operational alternative for preview rendering, with some unresolved issues.
```
Comment (ikesau) [inline adminSiteClient/GdocsPreviewPage.tsx:249]: Soft preference to not bother keeping this part in. 🙂 

I think either a) we'll know in the next month or so if this solution ends up not working for us (in which case you'll be able to reference this branch still) or b) we'll be fine with it and all these comments will just hang around without a clear expiry date.

File: adminSiteClient/GdocsPreviewPage.tsx [hunk 3]
```diff
@@ -233,9 +234,46 @@ export const GdocsPreviewPage = ({ match, history }: GdocsMatchProps) => {
...
+                {/* 
+                    This uses the full SSR rendering pipeline. It is more accurate but comes
+                    with an additional requests to the Google API and has a less polished
+                    authoring experience at the moment (content flashes and scrolling position
+                    resets on every change)
+                */}
+                <iframe
+                    src={`/gdocs/${gdoc.id}/preview#owid-article-root`}
+                    style={{ width: "100%", height: "inherit", border: "none" }}
+                    key={gdoc.revisionId}
```
Comment (ikesau) [inline adminSiteClient/GdocsPreviewPage.tsx:246]: Just checking: this is only needed so that the iframe knows when to refresh?

File: site/gdocs/BlockErrorBoundary.tsx [hunk 4]
```diff
@@ -1,48 +1,57 @@
...
-    ) : (
-        <>{children}</>
     )
 }
 
 export const BlockErrorFallback = ({
     error,
     resetErrorBoundary,
-    className = "",
+    className = getLayout(),
```
Comment (ikesau) [inline site/gdocs/BlockErrorBoundary.tsx:21]: What's the motivation for this change? It's better than `""` I suppose, but calling `getLayout` like this will _always_ return the `layouts[default][default]` layout, which means that if it's in a column that's less than 12 wide, it'll break the layout.

**Review by ikesau — APPROVED** [review]
Yup! Code's good for me too, I tested it locally but would also like to take it for a spin on a staging server to see how the error handling goes.

**Comment by mlbrgl** [comment]
@danyx23 @ikesau thanks for the reviews!

- Added missing nginx configuration to the PR description above and updated configuration on `snow`.
- Testing URLs:
  - google doc: https://docs.google.com/document/d/1gtECVQfzPCyZ5xGMXce5ONGitTQlLT3GnuXbrNWH2Ss/edit#
  - admin preview: https://snow.owid.cloud/admin/gdocs/1gtECVQfzPCyZ5xGMXce5ONGitTQlLT3GnuXbrNWH2Ss/preview
  - baked: https://snow-owid.netlify.app/test-error
  - public preview: https://snow.owid.cloud/gdocs/1gtECVQfzPCyZ5xGMXce5ONGitTQlLT3GnuXbrNWH2Ss/preview

File: adminSiteClient/GdocsPreviewPage.tsx (hunk 3 above)
Comment (mlbrgl) [reply adminSiteClient/GdocsPreviewPage.tsx:246]: yes, exactly

**Review by danyx23 — APPROVED** [review]
Thanks for fixing the deploy on snow - LGTM!

File: site/gdocs/BlockErrorBoundary.tsx (hunk 4 above)
Comment (mlbrgl) [reply site/gdocs/BlockErrorBoundary.tsx:21]: Thanks! Yes the error was a bit squished in e.g. sticky layouts since it would start at col 5 (and stop at the edge of the grid). This might not be entirely obvious but this only affects client-side errors - so testing this requires throwing on the client only (e.g. `if (typeof window !== "undefined") throw new Error("This is a test error")` in Chart.tsx)

**Comment by mlbrgl** [comment]
Redeploying on snow with the latest changes. Will push live tomorrow.

**Comment by mlbrgl** [comment]
thanks @danyx23!

@ikesau I integrated your feedback, it's being deployed on snow now. I noted you wanted to check the error handling further. Please let me know when you get a chance to do it, I'll wait for your confirmation before going live.

**Comment by ikesau** [comment]
:shipit:
